src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,304 @@
+            ISymbol symbol = context.Symbol;
+
+            // Don't report diagnostics on symbols that are marked as Preview
+            if (ProcessPreviewAttribute(symbol, requiresPreviewFeaturesSymbols))
+            {
+                return;
+            }
+
+            var reportDiagnostic = symbol switch
+            {
+                ITypeSymbol typeSymbol => TypeSymbolHasAttribute(typeSymbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType),
+                IPropertySymbol or IMethodSymbol => PropertyOrMethodSymbolHasAttribute(symbol, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType),
+                _ => false
+            };
+
+            if (reportDiagnostic)
+            {
+                if (requiresPreviewFeaturesSymbolsToUsageType.TryGetValue(symbol, out PreviewFeatureUsageType usageType))
+                {
+                    if (usageType == PreviewFeatureUsageType.StaticAbstract)
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(StaticAbstractIsPreviewFeatureRule));
+                    }
+                    else if (usageType == PreviewFeatureUsageType.EmptyInterface)
+                    {
+                        context.ReportDiagnostic(symbol.CreateDiagnostic(ImplementsEmptyPreviewInterfaceRule, symbol.Name));
+                    }
+                }
+                else
+                {
+                    context.ReportDiagnostic(symbol.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+                }
+            }
+        }
+
+        private static void BuildSymbolInformationFromOperations(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            if (OperationUsesPreviewFeatures(context, requiresPreviewFeaturesSymbols, out ISymbol? symbol))
+            {
+                IOperation operation = context.Operation;
+                if (operation is ICatchClauseOperation catchClauseOperation)
+                {
+                    operation = catchClauseOperation.ExceptionDeclarationOrExpression;
+                }
+
+                context.ReportDiagnostic(operation.CreateDiagnostic(GeneralPreviewFeatureAttributeRule, symbol.Name));
+            }
+        }
+
+        private static bool OperationUsesPreviewFeatures(OperationAnalysisContext context, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, [NotNullWhen(true)] out ISymbol? symbol)
+        {
+            IOperation operation = context.Operation;
+            if (operation is IInvocationOperation || operation is IPropertyReferenceOperation)
+            {
+                ISymbol containingSymbol = context.ContainingSymbol;
+                if (OperationUsesPreviewFeatures(containingSymbol, requiresPreviewFeaturesSymbols) || OperationUsesPreviewFeatures(containingSymbol.ContainingSymbol, requiresPreviewFeaturesSymbols))
```

Comment:
Any reason we are only looking up one level of `.ContainingSymbol` and not walking the entire `.ContainingSymbol` chain?

Reply:
No reason. The symbol analysis stops checking after 1 level, so I decided to do the same thing here for now. I can certainly see needing to check more than 1 level up for nested Preview method calls. Once I have more feedback from Tanner/Immo, I'll change/keep the logic here.